### ISSUE TITLE
Consider using ahash with a seed

### ISSUE THREAD

**Issue by ritchie46** [simd-lite/simd-json#290]
Hi, I read the `known_key` optimization, using fxhash because the hash is stable.

A `RandomState::with_seeds` would also produce stability, but favor the faster hash function and better hash quality. I think this can improve performance.

We especially noticed in polars that performance of fxhash can drop of a cliff on some data patterns (factors of 2 for instance).  https://github.com/tkaitchuck/aHash/blob/master/compare/readme.md#fxhash

Any thoughts?

**Comment by Licenser** [comment]
Hi,

I like the idea. That'd be a nice improvement, if nothing else, something we could put behind a feature flag to let people choose what they want.

One thing that brings up is how to pick a seed, it could be pre-randomized on every startup with something like a lazy static, or settable by the user in a global variable. Since you've a usecase right now, what would work best for you there?

**Comment by ritchie46** [comment]
Assuming you don't want the `RandomState` in the public API. You can pick a fixed seed on startup with lazy static/ once cell. Or even simpler is a lazy OnceCell with a random initialized `RandomState` in it. The size of the `RandomState` is only 4 `u64`, so no problem keeping that around.

https://rust-lang.github.io/hashbrown/src/ahash/random_state.rs.html#88-189

**Comment by Licenser** [comment]
This is quite a bit more complicated than anticipated, AHasher doesn't seem to have an interface that allows initializing it with either a given RandomSeed, nor one that allows initializing it with keys:

https://docs.rs/ahash/0.8.3/ahash/struct.AHasher.html

**Comment by ritchie46** [comment]
No, it is the other way around. A `RandomState` produces a stable `AHasher`. 

`RandomState` implements the [BuildHasher](https://doc.rust-lang.org/nightly/core/hash/trait.BuildHasher.html) trait. So calling `RandomState::build_hasher()` returns an `AHasher`.

Here is the source to confirm: https://rust-lang.github.io/hashbrown/src/ahash/random_state.rs.html#198-233

**Comment by Licenser** [comment]
I know, `RandomState` is the Builder and gets passed as a generic type to the HashMap. So we'd have to make a new struct I snarkily called it `NotSoRandomState` that acts as a `BuildHasher` and generates an `AHasher` the problem is that the hash builder has to call `AHasher::from_random_state` and that's a private function.

**Comment by ritchie46** [comment]
Have you got a small snippet that shows the problem?

I believe we can store the Randomstate in a static and clone it when we need it to instantiate a HashMap. 

We so something similar in polars.

**Comment by Licenser** [comment]
Heya,

okay so first of all, you're right it's possible and simple, I just was too dens yesterday evening to work it out. Talking it through helped. Let me walk you through my thinking perhaps there are still better ways I missed.

I suspect it's partially that the solution needed for a library crate is a bit different than for an application crate. Replacing all `HashMap::new()` works well inside an application where you have complete control over the interface. in simd-json that won't work  as we'd run the risk that someone calls `Value::from(HashMap::new())` (or equivalent) that then will insert a hashmap with a different hasher causing the known key to quietly not work for that specific Object. That'd be bad :tm:.

So the solution I aimed for was making a new hasher that effectively works the same as RandomState just that instead of default initializing a random state, it always re-uses the same one. That way the type ensures that the correct hasher is used and users can't accidentally break the interface by creating a HashMap with the wrong hasher.

**Comment by Licenser** [comment]
Reopening this as I had to yank 0.10, the changes were breaking the API in a way that's unreasonable so I'll have to work that out first.

**Comment by Licenser** [comment]
so `HashMap::new()` (for all the hashmaps) is only defined for the `DefaultHashBuilder`, so by introducing the `ObjectHasher`, `HashMap:new()` no longer exists unless `ObjectHahser = DefaultHashBuilder`. This is the way it works for all hash maps and the "workaround" is to use `HashMap::with_hahser(ObjectHahser::default())`.

I'm going through some examples to see if this is a reasonable tradeoff, I'd have loved to keep the ability to use `::new()` as this changes quite a few things, OTOH it's a breaking release so that might be fair. But before making a final call I want to get an impression of what the impact is and think about if there is an alternative to get both `::new()` and `ObjectHasher` but I'm not hopeful on that front given the limitations of rusts type system in that area

**Comment by ritchie46** [comment]
You could add a trait that implements it:

```rust
/// Representation of a JSON object
pub type Object = HashMap<String, Value, ObjectHasher>;

pub trait ObjectInit {
    fn new() -> Self;
}

impl ObjectInit for Object {
    fn new() -> Self {
        Self::default()
    }
}
```

**Comment by Licenser** [comment]
That could work, but I think that would lead to `new()` not be easily usable when the `ObjectHahser` is `DefaultHashBuilder`, as then the `new()` function can't be identified clearly

**Comment by ritchie46** [comment]
> That could work, but I think that would lead to `new()` not be easily usable when the `ObjectHahser` is `DefaultHashBuilder`, as then the `new()` function can't be identified clearly

In that case that trait should not be compiled. Can add a `#[cfg(..)]` on top.

**Comment by Licenser** [comment]
that's an itneresting idea, let me try that!

**Comment by Licenser** [comment]
I like it, thank you! That is quite elegant I didn't think of that :) seems to work well.

**Comment by ritchie46** [comment]
Nice! :)

**Comment by Licenser** [comment]
okay worked this out kinks are solved, released as 0.10.2
